File: InterviewPrep/QueueUsingTwoStacks.cs
using System;
using System.Collections.Generic;
using System.Linq;

class QueueUsingStacks<T>
{
    private Stack<T> _inStack;
    private Stack<T> _outStack;

    public QueueUsingStacks()
    {
        _inStack = new Stack<T>();
        _outStack = new Stack<T>();
    }

    public void Enqueue(T item)
    {
        _inStack.Push(item);
    }

    public T Dequeue()
    {
        if (_outStack.Count == 0)
            MoveStacks(_inStack, _outStack);

        return _outStack.Pop();
    }

    public void Front()
    {
        if (_outStack.Count == 0)
            MoveStacks(_inStack, _outStack);

        Console.WriteLine(_outStack.Peek());
    }

    private void MoveStacks(Stack<T> from, Stack<T> to)
    {
        while (from.Count > 0)
            to.Push(from.Pop());
    }
}

class Solution
{
    static int Main(string[] args)
    {
        int queries = Convert.ToInt32(Console.ReadLine().Trim());
        QueueUsingStacks<int> theQueue = new QueueUsingStacks<int>();

        for (int i = 0; i < queries; i++)
        {
            List<int> query = Console.ReadLine()
                                     .Trim()
                                     .Split(' ')
                                     .Select(val => Convert.ToInt32(val))
                                     .ToList();

            // The problem's description guarantees we won't get a bad query index,
            // and that all operations are valid.

            switch (query[0])
            {
                case 1:
                    theQueue.Enqueue(query[1]);
                    break;

                case 2:
                    theQueue.Dequeue();
                    break;

                case 3:
                    theQueue.Front();
                    break;
            }
        }
        return 0;
    }
}

File: Hackerrank/Medium/QueueUsingTwoStacks.swift
import Foundation

class QueueUsingStacks<T>
{
    private var enter: [T] = []
    private var exit: [T] = []

    public func enqueue(item: T) -> Void
    {
        enter.append(item)
    }

    @discardableResult public func dequeue() -> T
    {
        if (exit.isEmpty)
        {
            moveStacks()
        }

        return exit.removeLast()
    }

    public func printFront() -> Void
    {
        if (exit.isEmpty)
        {
            moveStacks()
        }

        print(exit.last!)
    }

    private func moveStacks() -> Void
    {
        while (!enter.isEmpty)
        {
            exit.append(enter.removeLast())
        }
    }
}

func Main()
{
    guard let queries = Int((readLine()?.trimmingCharacters(
                               in: .whitespacesAndNewlines))!)
    else { fatalError("Bad input") }

    let theQueue = QueueUsingStacks<Int>()

    for _ in 1...queries
    {
        let query = readLine()?.trimmingCharacters(in: .whitespaces)
                               .split(separator: " ")
                               .compactMap { Int($0) }

        // The problem's description guarantees we won't get a bad query index,
        // and that all operations are valid.

        switch query![0]
        {
        case 1:
            theQueue.enqueue(item: query![1])

        case 2:
            theQueue.dequeue()

        case 3:
            theQueue.printFront()

        default:
            print("We should not run this clause.")
        }
    }
}

Main()
